parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -212,8 +212,20 @@ private static MetadataFilter filter(boolean skipRowGroups) {
   }
 
   private static <T> List<T> runAllInParallel(int parallelism, List<Callable<T>> toRun) throws ExecutionException {
-    LOG.info("Initiating action with parallelism: {}", parallelism);
-    ExecutorService threadPool = Executors.newFixedThreadPool(parallelism);
+    if (toRun.isEmpty()) {
+      return Collections.emptyList();
+    } else if (toRun.size() == 1) {
+      // Prevent instantiating a ThreadPool if only one task
+      try {
+		return Collections.singletonList(toRun.get(0).call());
+      } catch (Exception e) {
+        throw new ExecutionException(e);
+      }
+    }
+    // Prevent instantiating a pool with many threads if not useful
+    int limitedParallelism = Math.min(parallelism, toRun.size());
```

Comment:
This is controlled by the parallelism parameter. If it is configured to parallelism=8 but the machine has 2 cores, then 8 threads may improve IO usage. I would not rely on _Runtime.getRuntime().availableProcessors()_ to override parallelism